feature(UI): Help modal form

PedroRamos360 opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #52223

<!-- Feel free to add any additional description of changes below this line -->

I read the suggestions in the issue #52223 and tried to create a first version of the modal. Lots of the comments have diversions but I built the form in a way that makes sense to me, I am open to suggestions and implementing changes as well as refactoring the code to best meet the guidelines.

Here are some pictures of what the form looks like:
After clicking on the help button:
![Screenshot from 2024-02-21 12-34-24](https://github.com/freeCodeCamp/freeCodeCamp/assets/53490820/3b3b6ffd-e5bc-45f5-b7a9-8d43a5af8c45)

After clicking on create help post on the forum:
Submit is active (user clicked in both checkboxes and typed at least 100 characters):
![Screenshot from 2024-02-21 12-35-18](https://github.com/freeCodeCamp/freeCodeCamp/assets/53490820/106287be-6ccd-442f-b0b1-318f0699365c)
Submit not active
![Screenshot from 2024-02-21 12-34-30](https://github.com/freeCodeCamp/freeCodeCamp/assets/53490820/da6fad63-2ae2-4f23-8073-468048b145c8)

After clicking on the submit button the user is redirected to the page to create a help post with his description automatically filled in.

Building this feature I had some doubts about the free code camp code architecture that I would appreciate if anyone could help me with such as:
- How is the process to implementing new translations in the codebase?
- What components are the standard to use? The form itself I didn't know what to use, I saw some places where it was being used normal forms others imported from libraries.
- Is it okay to style I component directly using style tag? Should I create a .css file to style each component or does the application uses bootstrap or some other styling framework?

PedroRamos360 committed:
feature(UI): Help modal form

PedroRamos360 committed:
Merge branch 'main' into ask-for-help-form

naomi-lgbt commented:
I'm not sure we should lose the two links in the process.

ilenia-magoni commented:
> * How is the process to implementing new translations in the codebase?

The strings are added to translations.json, and accessed with the `t` function

PedroRamos360 commented:
> I'm not sure we should lose the two links in the process.

What do you mean losing the two links? I didn't change the original modal, I just changed the functionality of the "create a help post on the forum" button to show the form instead of directly redirecting the user to the forum.

PedroRamos360 commented:
> > * How is the process to implementing new translations in the codebase?
> 
> The strings are added to translations.json, and accessed with the `t` function
> 
> Learn more: https://contribute.freecodecamp.org/#/how-to-work-on-localized-client-webapp?id=how-to-work-on-localized-client-webapp

Thank you! I am going to read that and try to implement the translations now.

PedroRamos360 committed:
Update test cases to match new page design

PedroRamos360 committed:
Fix test cases not finding the checkboxes

PedroRamos360 committed:
Remove unecessary translations of other languages

PedroRamos360 committed:
Merge branch 'main' into ask-for-help-form

raisedadead commented:
> What do you mean losing the two links?

The modal code doesn't seem to use any hyperlinks to the web pages to the "Read-Search-Ask" and "already answered...". They are just text checkboxes. People need to be able to click a link and visit the pages and then confirm they have actually read the documentation.

PedroRamos360 committed:
Merge branch 'main' into ask-for-help-form

PedroRamos360 committed:
Associate the checkboxes with their links

raisedadead commented:
> The links were already been shown in the first part of the modal...

Thanks - I missed that. In any case it is better and somewhat expected that the links be in the checkboxes too. 

This is looking good. Some one will have this reviewed soon. Thanks a lot for working on this.

Sembauke commented:
Thank you for this pull-request @PedroRamos360 
I am having some issues loading the link text though even after running `clean:client`:
<img width="603" alt="image" src="https://github.com/freeCodeCamp/freeCodeCamp/assets/46919888/92a2e811-295d-4eaf-9dfa-1f34091637a1">
Perhaps I am missing something quite obvious?

Sembauke commented on client/src/templates/Challenges/components/help-modal.tsx:
```diff
@@ -46,6 +46,38 @@ const generateSearchLink = (title: string, block: string) => {
+          target='_blank'
+          title={t(i18nkey)}
+        >
+          placeholder
```
```suggestion
          {t(i18nkey)}
```

Sembauke commented on client/i18n/locales/english/translations.json:
```diff
@@ -383,6 +384,8 @@
     "percent-complete": "{{percent}}% complete",
     "project-complete": "Completed {{completedChallengesInBlock}} of {{totalChallengesInBlock}} certification projects",
     "tried-rsa": "If you've already tried the <0>Read-Search-Ask</0> method, then you can ask for help on the freeCodeCamp forum.",
+    "read-search-ask-checkbox": "<0>I have tried the Read-Search-Ask method</0>",
+    "similar-questions-checkbox": "<0>I searched for similar questions that have already been answered on the forum</0>",
```
```suggestion
    "read-search-ask-checkbox": "I have tried the Read-Search-Ask method",
    "similar-questions-checkbox": "I searched for similar questions that have already been answered on the forum",
```

Sembauke reviewed (changes requested):
also should align the text left in `help-modal.css`:

```
a {
text-align: left;
}
```

PedroRamos360 committed:
Apply suggested changes for review

PedroRamos360 commented:
@Sembauke Thanks for the review! I didn't know what command to run to apply the translations before that is why I didn't catch that error. Now I ran it and applied your suggested changes.

Sembauke reviewed (changes requested):
<img width="598" alt="image" src="https://github.com/freeCodeCamp/freeCodeCamp/assets/46919888/f1408dee-e693-4f01-8387-0624a0325484">

It seems like the submit button is still inactive but clickable after checking all the boxes.

Sembauke commented:
We can change this in `onSubmit` with a conditional. 

We should also let the user know why it is still disabled after adding text. (If the text is too short)

ilenia-magoni commented:
What happens if the What's happening message + code exceede the max of characters that can be imported as starting message in the forum?
At that point it doesn't even import the text of the box, see trying to open an help post from RPG step 165

![image](https://github.com/freeCodeCamp/freeCodeCamp/assets/26656284/497adf15-7e45-4276-b3a6-927206cf3bcf)


```
### Tell us what's happening:

### Your code so far



WARNING

The challenge seed code and/or your solution exceeded the maximum length we can port over from the challenge.

You will need to take an additional step here so the code you wrote presents in an easy to read format.

Please copy/paste all the editor code showing in the challenge from where you just linked.

\`\`\`text
Replace these two sentences with your copied code.
Please leave the ``` line above and the ``` line below,
because they allow your code to properly format in the post.
\`\`\`

### Your browser information:

User Agent is: <code>Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36</code>

### Challenge Information:
Learn Basic JavaScript by Building a Role Playing Game - Step 165
http://localhost:8000/learn/javascript-algorithms-and-data-structures-v8/learn-basic-javascript-by-building-a-role-playing-game/step-165
```

It does the same for a very long tell us what's happening message, meaning the camper could loose the question they asked.

naomi-lgbt commented:
Likely need to update the help saga to parse this new form field as the question.

PedroRamos360 commented:
> It seems like the submit button is still inactive but clickable after checking all the boxes.

I also included a verification for the input need to have at least 100 characters, I can remove this verification or just display this information to avoid people being confused. It doesn't need to be 100 characters necessarily but I believe a character limit makes sense otherwise they could just send an empty question to the forum.

PedroRamos360 commented:
> What happens if the What's happening message + code exceede the max of characters that can be imported as starting message in the forum?
> At that point it doesn't even import the text of the box, see trying to open an help post from RPG step 165

When I tested the message import was working, I am going to review it later to see if I missed something. About the max of characters problem I could put a limit based on the length of the code the user wrote or just a fixed limit for the input. I don't know how long is the character limit but if is high I believe the fixed limit would work just fine.

ilenia-magoni commented:
It's working if the URI is shorter than the limit
https://github.com/freeCodeCamp/freeCodeCamp/blob/80fb22c36c6513889cf4b44c13995c55005eff72/client/src/templates/Challenges/redux/create-question-epic.js#L164
If it's longer the written question is lost

A static limit is not going to work, as the limit is the URI length, not the code or the question length, but both together plus the other frills

PedroRamos360 committed:
Added text limit and replaced textarea for the FormControl component

PedroRamos360 commented:
> It's working if the URI is shorter than the limit
> 
> https://github.com/freeCodeCamp/freeCodeCamp/blob/80fb22c36c6513889cf4b44c13995c55005eff72/client/src/templates/Challenges/redux/create-question-epic.js#L164
> 
> 
> If it's longer the written question is lost
> A static limit is not going to work, as the limit is the URI length, not the code or the question length, but both together plus the other frills

I updated the code now to add a limit of 500 characters for the textarea. The total limit of 8000 characters is a really big limit, the 500 characters limit for the textarea should be more than enough. Also even if someone manages to pass that limit (that would probably happen only if trying do it on purpose) there is already treatment for this case at the line you reffered to in which the code uses an alternative URI:
https://github.com/freeCodeCamp/freeCodeCamp/blob/80fb22c36c6513889cf4b44c13995c55005eff72/client/src/templates/Challenges/redux/create-question-epic.js#L164

Also, blocking the person from sending the message if the total limit was surpassed would not solve the problem as the code itself could have more than 8000 characters and it wouldn't help making their description smaller. In that case the person would be completely blocked from sending a question because her description would be bigger than the limit and smaller than the limit always.

PedroRamos360 committed:
Merge branch 'main' into ask-for-help-form

ilenia-magoni commented:
> The total limit of 8000 characters is a really big limit, the 500 characters limit for the textarea should be more than enough.

Halfway through the RPG project the code alone makes it too big. Is the question kept if the code is too big? It can be disheartening if the camper takes time to fill a question and then the forum precompiled post doesn't include it.

It's not the scope of this issue/PR, but it could make sense if only the file that needs changing is ported over maybe, that would reduce more the issue with too long code

PedroRamos360 commented:
> > The total limit of 8000 characters is a really big limit, the 500 characters limit for the textarea should be more than enough.
> 
> Halfway through the RPG project the code alone makes it too big. Is the question kept if the code is too big? It can be disheartening if the camper takes time to fill a question and then the forum precompiled post doesn't include it.
> 
> It's not the scope of this issue/PR, but it could make sense if only the file that needs changing is ported over maybe, that would reduce more the issue with too long code

Yes, you are right, it would be really frustrating to have to write the question all over again. There are three solutions that I can think of: The best one that would work for any case that would be updating the https://forum.freecodecamp.org to receive the questions in some other form like creating a table for those questions and in the URI just sending the ID, but I believe that would be unecessarily complicated. Or I could simply check if the length of the URI will pass 8000 characters if the description is filled and then simply not display the description textarea. This way the user won't waste time writing a question that won't be saved and can write it when it gets to the page. And finally even simpler than all of that would simply be always remove the textarea, or replace it with an input field the user has to fill to prove they understand like saying "I read the articles in both checkboxes and understand the guidelines of creating a question in the forum" or something like that.

PedroRamos360 commented:
It would also be possible to just take the code out when the limit passes 8000 characters and just putting the description.

PedroRamos360 committed:
Remove useless executeGA and add description to altTextMessage as well

jeremylt commented:
You could also send the editable region with a check if there are changes outside of the editable region

a2937 commented on client/src/templates/Challenges/redux/create-question-epic.js:
```diff
@@ -139,7 +142,7 @@ function createQuestionEpic(action$, state$, { window }) {
       const addCodeOne = i18next.t('forum-help.add-code-one');
```
Hi there. Your implementation seems very nice so far. There is just one personal nitpick of mine. If it's not possible to grab the whole file, maybe we can grab just the edit region rather than force the camper to copy and paste the edit code? I can't see the edit region to any challenge being big enough to warrant things being simply to long. I don't believe campers appreciate having to post code when it doesn't seem necessary, and some have trouble keeping their formatting straight.

PedroRamos360 replied:
> Hi there. Your implementation seems very nice so far. There is just one personal nitpick of mine. If it's not possible to grab the whole file, maybe we can grab just the edit region rather than force the camper to copy and paste the edit code? I can't see the edit region to any challenge being big enough to warrant things being simply to long. I don't believe campers appreciate having to post code when it doesn't seem necessary, and some have trouble keeping their formatting straight.

You are right. It is better showing only the editable region than nothing. I added a third fallback URI that sends only the editable region to the forum post even then if somehow that URI is bigger then it will fallback to the altURI where the user will have copy and paste the code.

PedroRamos360 committed:
Add fallback URI with only the editable region code

Sembauke reviewed (changes requested):
Hey @PedroRamos360,
This is looking great! The only thing I would do differently is:

1. Make the character count dynamic e.g. : `Add a minimum of X more characters`
2. When the user has reached this: ` X characters left `
3. Position it at the bottom of the field with a top padding and alignment with the text above the input field.

Also: The submit button is still clickable when it is disabled.

PedroRamos360 commented:
> Hey @PedroRamos360, This is looking great! The only thing I would do differently is:
> 
> 1. Make the character count dynamic e.g. : `Add a minimum of X more characters`
> 2. When the user has reached this: `X characters left`
> 3. Position it at the bottom of the field with a top padding and alignment with the text above the input field.
> 
> Also: The submit button is still clickable when it is disabled.

Do you mean like showing this text `Add a minimum of X more characters` when user didn't type anything and then show this `X characters left` when they start typing?

Sembauke commented:
Before the minimum is reached:
`Add a minimum of X more characters`

Then:
`X characters left`

raisedadead commented:
Does it have to part of this PR? I would like to limit the scope of the PR because it is becoming increasingly complex?

Sembauke commented:
Let me know @PedroRamos360, feel free to add this in another pr!

PedroRamos360 committed:
Add counter with characters left and reset form values when user closes the modal

Sembauke commented:
Awesome, this is exactly what I meant. 

The issue with the submit button still exists though!

lasjorg commented:
It looks good.

I might suggest adding a little more vertical separation between the elements inside the form. Like a `row-gap` of `0.5rem` to the flexbox container inside the form.

Can we get @bbsmooth to OK the accessibility if they have the time?

---

Maybe we should stop the back button navigation and warn the user if the `textarea` has been filled in. Just so people do not write a bunch and hit the back button by accident.

It would be nice if the form remembered what was typed but that might be asking too much.

PedroRamos360 commented:
> Awesome, this is exactly what I meant.
> 
> The issue with the submit button still exists though!

Are you sure? I fixed that in the previous commit I extracted the verification to the disabled button to a function then used that same function in the form submit function, I tested again now and it is not clickable, before I could click on it by only filling the 50 characters and not checking the checkboxes but now the only way is by filling the 50 characters and clicking the checkboxes.

PedroRamos360 commented:
> It looks good.
> 
> I might suggest adding a little more vertical separation between the elements inside the form. Like a `row-gap` of `0.5rem` to the flexbox container inside the form.
> 
> Can we get @bbsmooth to OK the accessibility if they have the time?
> 
> Maybe we should stop the back button navigation and warn the user if the `textarea` has been filled in. Just so people do not write a bunch and hit the back button by accident.
> 
> It would be nice if the form remembered what was typed but that might be asking too much.

Actually, it was that way before it did not erased the content even if you closed the modal and opened it again but that started creating some problems in the interface and I decided to reset all the states every time the user gets out of the page. I think I can fix that, I am going to just avoid reseting the values when the person clicks the back button.

PedroRamos360 committed:
Keep form values after clicking the back button

lasjorg commented:
Do the input elements not need labels or aria attributes?

Or is the link text and title enough and if so do the input elements need to be associated with the links using `aria-labelledby`?

bbsmooth commented:
> Can we get @bbsmooth to OK the accessibility if they have the time?

Definitely. I have been watching this PR from a distance and it sounds like it is about ready to go. I will make a point to test it by the end of this coming weekend.

bbsmooth commented:
There are quite a few accessibility issues. Some of them will be relatively easy to fix and I'll start adding code suggestions. A few are going to be much more complicated, such as the textarea with the maximum character limit that dynamically updates how many characters you have left, and we may decide to save those for separate PRs.

bbsmooth commented on client/src/templates/Challenges/components/help-modal.tsx:
```diff
@@ -53,61 +87,198 @@ function HelpModal({
+          </form>
+        ) : (
+          <>
+            <h3 className='help-modal-heading'>
+              <Trans i18nKey='learn.tried-rsa'>
+                <a
+                  href={RSA}
+                  rel='noopener noreferrer'
+                  target='_blank'
+                  title={t('learn.rsa')}
+                >
+                  placeholder
+                </a>
+              </Trans>
+            </h3>
```
```suggestion
            <p className='help-modal-heading'>
              <Trans i18nKey='learn.tried-rsa'>
                <a
                  href={RSA}
                  rel='noopener noreferrer'
                  target='_blank'
                  title={t('learn.rsa')}
                >
                  placeholder
                </a>
              </Trans>
            </p>
```

This sentence is not acting as a heading, it is just regular content.

bbsmooth commented on client/i18n/locales/english/translations.json:
```diff
@@ -383,6 +384,10 @@
+    "minimum-characters": "Add a minimum of {{characters}} more characters",
+    "characters-left": "{{characters}} characters left",
     "rsa": "Read, search, ask",
     "rsa-forum": "<strong>Before making a new post</strong> please see if your question has <0>already been answered on the forum</0>.",
```
```suggestion
    "rsa-forum": "<strong>Before making a new post</strong> please <0>check if your question has already been answered on the forum</0>.",
```

I would extend the link text so that it makes more sense to a screen reader user listening to just the link text.

bbsmooth commented:
Question. Do we really need a "Back" button on the second dialog that takes you back to the first dialog? The first dialog has the exact same information on it as the second dialog. I'm not sure I see how it helps to go back to the first dialog from the second. It seems to me if you are on the second dialog and decide you don't want to create a Help post then you just want to cancel completely, so there is no reason to go back to the first dialog. I recommend we replace the Back button on the second dialog with Cancel, just like it works on the first dialog.

bbsmooth commented on client/src/templates/Challenges/components/help-modal.tsx:
```diff
@@ -45,6 +45,40 @@ const generateSearchLink = (title: string, block: string) => {
+}: CheckbockProps) {
+  return (
+    <div style={{ display: 'flex', flexDirection: 'row' }}>
+      <input name={name} type='checkbox' onChange={onChange} checked={value} />
```
```suggestion
      <input name={name} type='checkbox' onChange={onChange} checked={value} aria-label={t(i18nkey)} />
```

An input always needs a programmatic name.

bbsmooth commented on client/src/templates/Challenges/components/help-modal.tsx:
```diff
@@ -45,6 +45,40 @@ const generateSearchLink = (title: string, block: string) => {
+          target='_blank'
+          title={t(i18nkey)}
+        >
+          {t(i18nkey)}
```
```suggestion
          {t(i18nkey)} <span className='sr-only'>{t('aria.opens-new-window')}
```

I think we need to let screen reader users know that these links open a new tab/window.

bbsmooth commented on client/src/templates/Challenges/components/help-modal.tsx:
```diff
@@ -45,6 +45,40 @@ const generateSearchLink = (title: string, block: string) => {
+          href={href}
+          rel='noopener noreferrer'
+          target='_blank'
+          title={t(i18nkey)}
```
```suggestion
```

We don't need the title attribute on text links. It is redundant and causes duplication for screen reader users.

bbsmooth commented on client/src/templates/Challenges/components/help-modal.tsx:
```diff
@@ -53,61 +87,198 @@ function HelpModal({
+                href={RSA}
+                t={t}
+              />
+              <CheckboxHelpModal
```
```suggestion
              <Spacer size='small' />
              <CheckboxHelpModal
```

It would be nice to add a little more space between these checkboxes to increase the target size for touch devices.

bbsmooth commented on client/src/templates/Challenges/components/help-modal.tsx:
```diff
@@ -53,61 +87,198 @@ function HelpModal({
+  const [similarQuestionsCheckbox, setSimilarQuestionsCheckbox] =
+    useState(false);
+  const DESCRIPTION_MIN = 50;
+  const DESCRIPTION_MAX = 500;
```
```suggestion
  const DESCRIPTION_MAX = 500;
  const formRef = useRef<HTMLFormElement>(null);

  useEffect(() => {
    if (showHelpForm) {
      formRef.current?.querySelector('input')?.focus();
    }
  }, [showHelpForm]);
```

Currently, when the "Create a help post on the forum" button is clicked, focus is lost when the second dialog opens and thus screen reader users hear nothing. We must manually set the focus so that they get an indication that something has changed. I think the most logical place to put the focus is on the first checkbox.

If we continue to have the Back button on the second dialog then we will also need to manually set focus when they go back to the first dialog. But I'm not adding that logic now because I'm hoping we reconsider having the Back button.

bbsmooth commented on client/src/templates/Challenges/components/help-modal.tsx:
```diff
@@ -1,16 +1,16 @@
 import { Modal } from '@freecodecamp/react-bootstrap';
-import React from 'react';
+import { Button, FormControl } from '@freecodecamp/ui';
+import React, { useCallback, useState } from 'react';
```
```suggestion
import React, { useCallback, useState, useRef, useEffect } from 'react';
```

Need these so we can set focus on second dialog.

bbsmooth commented on client/src/templates/Challenges/components/help-modal.tsx:
```diff
@@ -53,61 +87,198 @@ function HelpModal({
+              resetFormValues();
+              createQuestion();
+              closeHelpModal();
+            }}
```
```suggestion
            }}
            ref={formRef}
```

Needed so we can set focus on second dialog.

bbsmooth commented on client/src/templates/Challenges/components/help-modal.tsx:
```diff
@@ -53,61 +87,198 @@ function HelpModal({
-          <p>
-            <Trans i18nKey='learn.rsa-forum'>
-              <a
+              <CheckboxHelpModal
```
```suggestion
              <fieldset>
                <legend>You must confirm the following statements before you can submit your post to the forum.</legend>
              <CheckboxHelpModal
```

I'm adding content here so I understand this suggestion is perhaps slightly controversial. Besides the fact that these instructions will make it clear that both checkboxes must be checked in order to submit, I think even more importantly it will help screen reader users understand why their focus has shifted to the first checkbox after clicking the "Create a help post on the forum" button on the previous dialog.

bbsmooth commented on client/src/templates/Challenges/components/help-modal.tsx:
```diff
@@ -53,61 +87,198 @@ function HelpModal({
-          {t('buttons.cancel')}
-        </Button>
+                t={t}
+              />
```
```suggestion
              />
              </fieldset>
```

bbsmooth commented on client/src/templates/Challenges/components/help-modal.tsx:
```diff
@@ -53,61 +87,198 @@ function HelpModal({
-        </Button>
+                t={t}
+              />
+              <label htmlFor='description'>
+                {t('forum-help.whats-happening')}
+              </label>
```
```suggestion
              <label htmlFor='help-modal-form-description'>
                {t('forum-help.whats-happening')}
                <span className='sr-only'>50 character minimum</span>
              </label>
```

Of course this needs to be translated, I'm just too lazy to do that.

PedroRamos360 commented:
> Question. Do we really need a "Back" button on the second dialog that takes you back to the first dialog? The first dialog has the exact same information on it as the second dialog. I'm not sure I see how it helps to go back to the first dialog from the second. It seems to me if you are on the second dialog and decide you don't want to create a Help post then you just want to cancel completely, so there is no reason to go back to the first dialog. I recommend we replace the Back button on the second dialog with Cancel, just like it works on the first dialog.
> 
> Related to this. Is there a reason we need two separate dialogs for this? I think it would make accessibility better if we just had one.

About the back button, I agree, a Cancel button makes more sense, but I think it is good to keep the 2 part form although it is redudant. The goal of this feature, from what I understand, is to make it harder for the user to create a question in the forum to incentivate people that aren't really engaged with making a question and just want someone to tell them the answer to give up. Also I believe it can help users understand the seriousness of the alert.

PedroRamos360 committed:
Update client/src/templates/Challenges/components/help-modal.tsx

Apply suggested changes by bbsmooth

Co-authored-by: Bruce Blaser <bbsmooth@gmail.com>

PedroRamos360 committed:
Update client/src/templates/Challenges/components/help-modal.tsx

Apply bbsmooth suggestions

Co-authored-by: Bruce Blaser <bbsmooth@gmail.com>

PedroRamos360 committed:
Update client/i18n/locales/english/translations.json

Apply bbsmooth suggestions

Co-authored-by: Bruce Blaser <bbsmooth@gmail.com>

PedroRamos360 committed:
Update client/src/templates/Challenges/components/help-modal.tsx

Co-authored-by: Bruce Blaser <bbsmooth@gmail.com>

PedroRamos360 committed:
Update client/src/templates/Challenges/components/help-modal.tsx

Co-authored-by: Bruce Blaser <bbsmooth@gmail.com>

PedroRamos360 committed:
Update client/src/templates/Challenges/components/help-modal.tsx

Co-authored-by: Bruce Blaser <bbsmooth@gmail.com>

PedroRamos360 committed:
Update client/src/templates/Challenges/components/help-modal.tsx

Co-authored-by: Bruce Blaser <bbsmooth@gmail.com>

PedroRamos360 committed:
Update client/src/templates/Challenges/components/help-modal.tsx

Co-authored-by: Bruce Blaser <bbsmooth@gmail.com>

PedroRamos360 committed:
Apply suggestions from code review

Apply accessibility changes suggested by bbsmooth

Co-authored-by: Bruce Blaser <bbsmooth@gmail.com>

PedroRamos360 committed:
Fix compilation bugs and add translations

PedroRamos360 commented:
How can I make this <legend> text responsive to dark mode? If i just set the style: {color: white} it will will mess up light mode, is there a component for this in the project?
![image](https://github.com/freeCodeCamp/freeCodeCamp/assets/53490820/bbb62cfe-c95b-483c-92d6-7f36f8d04ee1)

lasjorg commented:
If you use the custom properties it should work.

https://github.com/freeCodeCamp/freeCodeCamp/blob/main/client/src/components/layouts/variables.css


We have Tailwind as well so it would be nice to switch whatever inline styles we can to Tailwind classes instead.

Unless there is a reason why we use inline styles. Not sure when/where we are using Tailwind or are supposed to use it.

---

Not sure what happened but the first modal page is missing spaces for the text when I looked.

![spaces](https://github.com/freeCodeCamp/freeCodeCamp/assets/28780271/e4436db4-08cf-419d-b596-89c1bcbbfbed)

bbsmooth commented on client/src/templates/Challenges/components/help-modal.tsx:
```diff
@@ -53,61 +89,208 @@ function HelpModal({
+          </form>
+        ) : (
+          <>
+            <p className='help-modal-heading'>
```
```suggestion
            <p>
```

It seems we don't need this class any  more since this content is just a regular `p` element now.

bbsmooth commented on client/src/templates/Challenges/components/help-modal.css:
```diff
@@ -13,8 +13,16 @@
+
+a {
+  text-align: left;
+}
```
```suggestion
}

.help-modal-body legend {
  color: var(--secondary-color);
}
```

Fix the added text above the checkboxes on the second dialog.

bbsmooth commented on client/src/templates/Challenges/components/help-modal.tsx:
```diff
@@ -53,61 +89,208 @@ function HelpModal({
+              block={true}
+              size='large'
+              variant='primary'
+              onClick={() => {
+                setShowHelpForm(false);
+              }}
+            >
+              {t('buttons.back')}
```
```suggestion
              onClick={closeHelpModal}
            >
              {t('buttons.cancel')}
```

Cancel button on second dialog instead of Back button.

bbsmooth commented on client/src/templates/Challenges/components/help-modal.tsx:
```diff
@@ -53,61 +89,208 @@ function HelpModal({
+              </fieldset>
+              <label htmlFor='help-modal-form-description'>
+                {t('forum-help.whats-happening')}
+                <span className='sr-only'>50 character minimum</span>
```
```suggestion
                <span className='sr-only'>{t('learn.min-50-max-500')}</span>
```

This needs to be translated.

bbsmooth commented on client/i18n/locales/english/translations.json:
```diff
@@ -383,8 +384,13 @@
+    "similar-questions-checkbox": "I searched for similar questions that have already been answered on the forum",
+    "minimum-characters": "Add a minimum of {{characters}} more characters",
+    "characters-left": "{{characters}} characters left",
+    "must-confirm-statements": "You must confirm the following statements before you can submit your post to the forum.",
```
```suggestion
    "must-confirm-statements": "You must confirm the following statements before you can submit your post to the forum.",
    "min-50-max-500": "50 character minimum, 500 character maximum",
```

PedroRamos360 committed:
Apply suggestions from code review

Co-authored-by: Bruce Blaser <bbsmooth@gmail.com>

PedroRamos360 committed:
Update client/src/templates/Challenges/components/help-modal.css

Co-authored-by: Bruce Blaser <bbsmooth@gmail.com>

PedroRamos360 committed:
Fix help-modal.spec.ts tests to match changes in help-modal.tsx

PedroRamos360 commented:
Commited the suggested changes and adapted the help-modal.spec.ts test to match the changes in help-modal.tsx but now some error is happening at test:curriculum that I don't understand, I am going to put more time figuring it out later but in the mean time if someone knows something about this error I would appreciate the help.

Sembauke commented:
This fail has likely nothing to do with your changes. I am re-running the CI to confirm. Otherwise I will look into this myself.

Sembauke commented:
Amazing! Thank you for your determination to this effort.

bbsmooth commented on client/src/templates/Challenges/components/help-modal.tsx:
```diff
@@ -53,61 +89,210 @@ function HelpModal({
+        resetFormValues();
+      }}
+      show={isOpen}
+    >
       <Modal.Header className='help-modal-header fcc-modal' closeButton={true}>
         <Modal.Title className='text-center'>
```
```suggestion
      aria-labelledby='ask-for-help-modal'
    >
      <Modal.Header className='help-modal-header fcc-modal' closeButton={true}>
        <Modal.Title id='ask-for-help-modal' className='text-center'>
```

I think we should give this dialog a programmatic (accessible) name. Currently, JAWS is just announcing "dialog". If we give the dialog a name, JAWS will announce "Ask for help dialog". I think this helps clarify that a dialog was opened specifically for the purpose of asking for help.

bbsmooth commented on client/src/templates/Challenges/components/help-modal.tsx:
```diff
@@ -45,6 +45,42 @@ const generateSearchLink = (title: string, block: string) => {
+        type='checkbox'
+        onChange={onChange}
+        checked={value}
+        aria-label={t(i18nkey)}
```
```suggestion
        aria-label={t(i18nkey)}
        aria-required='true'
```

You can't submit without checking these, so we should have `aria-required` on them.

bbsmooth commented on client/i18n/locales/english/translations.json:
```diff
@@ -95,7 +95,8 @@
     "confirm-amount": "Confirm amount",
-    "skip-advertisement": "Skip Advertisement"
+    "skip-advertisement": "Skip Advertisement",
+    "back": "Back"
```
```suggestion
    "back": "Back",
    "close-dialog": "Close dialog"
```

You might want to add this right after "close" on line 7? I just added it here since that's the only place github would let me add this suggestion in "buttons".

bbsmooth commented on client/src/templates/Challenges/components/help-modal.tsx:
```diff
@@ -53,61 +89,210 @@ function HelpModal({
+            </div>
+            <Spacer size='xxSmall' />
+
+            <Button
+              block={true}
+              size='large'
+              variant='primary'
+              type='submit'
+              disabled={!canSubmitForm()}
+            >
```
```suggestion
            <Button
              block={true}
              size='large'
              variant='primary'
              type='submit'
              disabled={!canSubmitForm()}
              {...(!canSubmitForm() && { tabIndex: '-1' })}
            >
```

If this button is truly disabled then it cannot accept keyboard focus.

bbsmooth commented:
Sorry, I added a few more accessibility updates, but I promise I'll stop now. There are still some accessibility issues that need to be addressed, but they are complicated enough that I'll save them for separate PRs.

PedroRamos360 committed:
Apply suggestions from code review

Co-authored-by: Bruce Blaser <bbsmooth@gmail.com>

PedroRamos360 committed:
Apply suggestions from code review

Co-authored-by: Bruce Blaser <bbsmooth@gmail.com>

PedroRamos360 committed:
Change tab index from string to number to fix typescript error

PedroRamos360 commented:
@bbsmooth thank you for your suggestions! Applied them now and tested and its working as expected. I didn't really understand the necessity of removing the tab focus on a disabled button for accessibility, does it have to deal with screen readers or something like that? Also if it is the most correct for a disabled button to not be focused maybe we could implement this directly in the Button component to avoid inconsistencies.

bbsmooth commented:
It's about consistency and expectations. If something can receive keyboard focus then the implication is that it is functional, and that might be confusing to some people. Also, the color contrast on disabled controls is allowed to go below minimum contrast requirements (which ours does), as long as the are truly disabled, which means no keyboard focus. So unless we want to make the text on that disabled Submit button a lot darker, then it can't allow keyboard focus.

lasjorg commented:
Do we need better visual feedback for the check box requirement? When filling out the text but forgetting to check the boxes it looks like the form is just not reacting to reaching the required text amount.

Should clicking outside the second modal screen close it now that it doesn't remember what you typed? There is some potential for accidentally losing what you typed with this approach.

PedroRamos360 commented:
> Do we need better visual feedback for the check box requirement? When filling out the text but forgetting to check the boxes it looks like the form is just not reacting to reaching the required text amount.
> 
> Should clicking outside the second modal screen close it now that it doesn't remember what you typed? There is some potential for accidentally losing what you typed with this approach.

What do you mean does not react to reaching the required text amount? If you mean that the submit button keeps disabled that is intentional as it is described above the checkboxes that it is necessary to check them so the user can continue.

lasjorg commented:
I'm saying if you forget to check the boxes and start typing and get "caught up" by the "enter this many characters" when you reach that and the form still can not be submitted it isn't super obvious what is missing.

Unlike the minimum characters requirement, there is no visual indication of the checkbox requirement.

bbsmooth commented:
> Do we need better visual feedback for the check box requirement?

Yes, the form needs better error handling in general. This is an issue I was going to work on in a separate PR. I didn't want to drag this one out any longer.

> Should clicking outside the second modal screen close it now that it doesn't remember what you typed?

That's a good point. Way to easy to accidentally click outside the box.

huyenltnguyen committed:
refactoring and minor fixes

huyenltnguyen committed:
Merge branch 'main' of https://github.com/freeCodeCamp/freeCodeCamp into pr/53824

huyenltnguyen committed:
tests

huyenltnguyen committed:
remove unused translation string

huyenltnguyen reviewed (approved):
@PedroRamos360 Thank you for the PR. And thank you everyone for the inputs and reviews.

To shorten the review loop, I've gone ahead and make some changes to the PR, as we both would like to have this feature in and to unblock some modal-related work.

<details>
  <summary>Screenshot of the modal after adjustments</summary>

<img width="629" alt="Screenshot 2024-03-25 at 12 40 49" src="https://github.com/freeCodeCamp/freeCodeCamp/assets/25715018/d72c14cd-6612-4789-ad02-127774b9446e">

</details>

Changes that worth mentioning are:
- Moving inline styles to the CSS file, and using class names as selectors instead of element names
- Wrapping the `input` elements for the checkbox inside a `label`, instead of adding `aria-label` to the inputs
- Allowing keyboard focus on the submit button, even if the button is disabled (I'll explain in a bit)
- Sending the description from the UI component to Redux via the `createQuestion` action, instead of retrieving the description using `document.getElementById()` within Redux
- Adding tests to cover the behavior of the submit button, as well as updating tests to use the `getByRole()` query

---

Regarding the submit button,

When an element has the `disabled` attribute set, the element doesn't receive focus or react to any interactions. This is bad UX when it comes to form. 

If a non-sighted user uses Tab to navigate through the form, they would only hear the inputs and the cancel button. The submit button, with `disabled` being `true`, is not announced as all, as if it is not in the DOM. This behavior could potentially confuse the user, as they don't know if there is a submit button on the page, let alone it being disabled.

I took this concern into account when I designed the shared `Button` component. The component accepts a `disabled` prop, but under the hood, it passes the value to the `aria-disabled`, rather than the `disabled` attribute.

This decision was also inspired by:
- https://css-tricks.com/making-disabled-buttons-more-inclusive/
- https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/pull/830 (the discussion is interesting to read, but if you need a TL;DR jump to https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/pull/830#pullrequestreview-880806333)

If this implementation results any a11y issues, that's an oversight on my part, and I'd like to address the issue on the library level instead of spot fixing in /learn.

That was to give you the context of the implementation. But we have this discussion because we need the disabled state, which is part of what's needed in order to remove Bootstrap. Once we are have total control over our UI and our UI component library, we can revisit the current UI/UX and adjust as needed.

IMHO, the submit button shouldn't be disabled as disabling it takes away our opportunity to guide the user or to give them feedback. But instead, we could enable the button at all time, but add some validation in its click handler function to block submission if the form is invalid, and display some error messages in a live region, telling the user what they need to change.

We are not in a position to invest in these UX improvements yet, but this topic has been living rent free in my head for quite a while.

raisedadead commented:
This is now live in production! Thanks @PedroRamos360 for working on this.

bbsmooth commented:
@huyenltnguyen 

> Wrapping the input elements for the checkbox inside a label, instead of adding aria-label to the inputs

We have links in the label text for the checkboxes (which is kind of an accessibility anti-pattern to begin with). Because those links open a new window, it is considered a best practice to let screen reader users know that, which is usually done with visually hidden "opens in new window" text immediately after the link. But we don't want screen reader users to hear that visually hidden text for the check box name itself, thus we were using `aria-label` for the checkbox to get around that.  So the trade-off here is that our code is a little cleaner but screen reader users aren't told that the link opens a new window. Not a show stopper, just wanted to make it clear why we were using `aria-label`. Personally, I would love to get rid of those links in the label text, especially since we have them already linked on the intro dialog page.

Concerning the `tabindex` issue on the disabled button, most accessibility professionals interpret WCAG to mean that that an element can only be considered disabled if it cannot be interacted with in any way, including having keyboard focus by tabbing to it. So removing `tabindex="-=1"` from the button does no longer makes it truly disabled and thus it must meet minimum contrast requirements to satisfy WCAG 1.4.3. (Yes, I am saying that the CSS Tricks article you linked to above is misinterpreting 1.4.3). So technically, the "disabled" Submit button in the dialog needs to have darker "Submit" text. But I wholeheartedly agree with you, we should not be disabling the Submit button in the first place and thus we won't need to worry about `tabindex` or contrast.

Also, I just checked production, and currently the Submit button is not disabled (even though it looks like it). I am able to click on it with both the keyboard and the mouse and get feedback from the form. The form is using the default error handling features build into the browser. They aren't perfect, but they will do for now. So I think it makes sense to get rid of `aria-disabled` on the Submit now.
